File: clojure/examples/basic_example.clj
;;;; ---------------------------------------------------------------------------
;;;;  Copyright (c) Microsoft Corporation. All rights reserved.
;;;; ---------------------------------------------------------------------------
;;;;
;;;; Basic example demonstrating the Copilot Clojure SDK.
;;;;
;;;; Run with:  clj -M:dev -m basic-example

(ns basic-example
  "Full working example showing client creation, tool definition,
  event subscription, and send-and-wait."
  (:require [copilot.client :as client]
            [copilot.session :as session]
            [copilot.define-tool :refer [deftool define-tool]]))

;; ============================================================================
;; 1. Define custom tools
;; ============================================================================

;; Option A: Using the deftool macro for concise definitions
(def facts
  {"javascript" "JavaScript was created in 10 days by Brendan Eich in 1995."
   "node"       "Node.js lets you run JavaScript outside the browser using the V8 engine."
   "clojure"    "Clojure was created by Rich Hickey and first released in 2007."})

(deftool lookup-fact
  "Returns a fun fact about a given topic."
  {:type       "object"
   :properties {:topic {:type        "string"
                        :description "Topic to look up (e.g. 'javascript', 'node', 'clojure')"}}
   :required   ["topic"]}
  [args _invocation]
  (let [topic (clojure.string/lower-case (or (:topic args) ""))]
    (get facts topic (str "No fact stored for " topic "."))))

;; Option B: Using define-tool function directly
(def get-time-tool
  (define-tool "get_current_time"
               "Returns the current server time."
               nil  ;; no parameters
               (fn [_args _invocation]
                 (str "Current time: " (java.time.Instant/now)))))

;; ============================================================================
;; 2. Main example flow
;; ============================================================================

(defn -main
  [& _args]
  (println "Starting Copilot SDK Example\n")

  ;; Create client - will auto-start CLI server
  (let [client (client/create-client :log-level "info")]
    (try
      ;; Create a session with our custom tools
      (let [sess (client/create-session! client
                   :tools [lookup-fact get-time-tool])]
        (println (str "Session created: " (session/session-id sess) "\n"))

        ;; Subscribe to all events
        (let [unsub (session/on! sess
                      (fn [event]
                        (println (str "  Event [" (:type event) "]: "
                                      (pr-str (:data event))))))]

          ;; Send a simple message and wait for response
          (println "Sending message: Tell me 2+2")
          (let [result (session/send-and-wait! sess {:prompt "Tell me 2+2"})]
            (println (str "\nResponse: "
                          (get-in result [:data :content])
                          "\n")))

          ;; Send a message that uses the lookup_fact tool
          (println "Sending message: Use lookup_fact to tell me about 'clojure'")
          (let [result (session/send-and-wait! sess
                         {:prompt "Use lookup_fact to tell me about 'clojure'"})]
            (println (str "\nResponse: "
                          (get-in result [:data :content])
                          "\n")))

          ;; --- v2.0 Features ---

          ;; Session Metadata
          (let [meta (client/get-session-metadata client (session/session-id sess))]
            (when meta
              (println "Session model:" (:model meta))))

          ;; Skills (uncomment to use)
          ;; (def skill-session (client/create-session! client
          ;;   {:skill-directories ["./skills"]
          ;;    :include-sub-agent-streaming-events true}))

          ;; Unsubscribe event handler
          (unsub))

        ;; Clean up session
        (session/destroy! sess))

      ;; Stop the client
      (let [errors (client/stop! client)]
        (when (seq errors)
          (println "Cleanup errors:" errors)))

      (println "Done!")

      (catch Exception e
        (println "Error:" (.getMessage e))
        (.printStackTrace e)
        (client/force-stop! client)))))

;; ============================================================================
;; 3. Advanced example: Permission handler & user input
;; ============================================================================

(comment
  ;; Example with permission handler and user input
  (let [client (client/create-client :log-level "info")]
    (let [sess (client/create-session! client
                 :tools [lookup-fact]
                 :on-permission-request
                 (fn [request _invocation]
                   (println "Permission requested:" (:kind request))
                   ;; Auto-approve all permissions
                   {:kind "approved"})

                 :on-user-input-request
                 (fn [request _invocation]
                   (println "Agent asks:" (:question request))
                   {:answer "yes" :wasFreeform true})

                 :hooks
                 {:on-pre-tool-use
                  (fn [input _invocation]
                    (println "About to run tool:" (:toolName input))
                    {:permissionDecision "allow"})

                  :on-post-tool-use
                  (fn [input _invocation]
                    (println "Tool completed:" (:toolName input))
                    nil)})]

      ;; Subscribe to specific event types
      (session/on-type! sess "assistant.message"
        (fn [event]
          (println "Assistant:" (get-in event [:data :content]))))

      (session/on-type! sess "session.idle"
        (fn [_event]
          (println "Session is idle")))

      (session/send-and-wait! sess {:prompt "Hello, world!"})

      (session/destroy! sess)
      (client/stop! client)))

  ;; Example: Listing models
  (let [client (client/create-client)]
    (client/start! client)
    (let [models (client/list-models! client)]
      (doseq [m models]
        (println (:id m) "-" (:name m))))
    (client/stop! client))

  ;; Example: Listing and resuming sessions
  (let [client (client/create-client)]
    (client/start! client)
    (let [sessions (client/list-sessions! client)]
      (doseq [s sessions]
        (println (:session-id s) "-" (:summary s)))
      (when-let [last-id (client/get-last-session-id! client)]
        (let [sess (client/resume-session! client last-id)]
          (session/send-and-wait! sess {:prompt "Continue where we left off"})
          (session/destroy! sess))))
    (client/stop! client))
  )

File: lua/examples/basic_example.lua
#!/usr/bin/env lua
--- Basic example demonstrating the Copilot Lua SDK.
--
-- This example:
--   1. Creates a CopilotClient and connects to the CLI server
--   2. Defines a custom tool (lookup_fact)
--   3. Creates a session with the tool registered
--   4. Sends a simple math question and prints the response
--   5. Sends a follow-up that uses the custom tool
--   6. Cleans up

-- Add the parent directory to the module path so we can find the copilot modules
local script_dir = arg[0]:match("(.-)[^/\\]+$") or "./"
package.path = script_dir .. "../?.lua;" .. script_dir .. "../?/init.lua;" .. package.path

local CopilotClient = require("copilot.client")
local define_tool   = require("copilot.define_tool")
local types         = require("copilot.types")

-- ---------------------------------------------------------------------------
-- Define a custom tool
-- ---------------------------------------------------------------------------

local facts = {
    javascript = "JavaScript was created in 10 days by Brendan Eich in 1995.",
    node       = "Node.js lets you run JavaScript outside the browser using the V8 engine.",
    lua        = "Lua was created in 1993 at PUC-Rio in Brazil and means 'moon' in Portuguese.",
    python     = "Python was created by Guido van Rossum and first released in 1991.",
}

local lookup_fact_tool = define_tool("lookup_fact", {
    description = "Returns a fun fact about a given programming topic.",
    parameters = {
        type = "object",
        properties = {
            topic = {
                type        = "string",
                description = "Topic to look up (e.g. 'javascript', 'lua', 'node')",
            },
        },
        required = { "topic" },
    },
    handler = function(args, invocation)
        local topic = (args.topic or ""):lower()
        local fact = facts[topic]
        if fact then
            return fact
        end
        return string.format("No fact stored for '%s'.", args.topic or "unknown")
    end,
})

-- ---------------------------------------------------------------------------
-- Permission handler (auto-approve everything for this example)
-- ---------------------------------------------------------------------------

local function permission_handler(request, invocation)
    print(string.format("[Permission] Session %s requested: %s",
        invocation.sessionId, request.kind or "unknown"))
    return types.PermissionRequestResult({ kind = "approved" })
end

-- ---------------------------------------------------------------------------
-- Main
-- ---------------------------------------------------------------------------

local function main()
    print("Starting Copilot SDK Lua Example")
    print(string.rep("-", 50))

    -- Create client (searches PATH for "copilot" CLI)
    local client = CopilotClient.new({
        logLevel = "info",
    })

    -- Start the client (spawns CLI server)
    local ok, err = client:start()
    if not ok then
        print("ERROR: Failed to start client: " .. tostring(err))
        os.exit(1)
    end
    print("Client started successfully.")

    -- List available models
    local models, models_err = client:list_models()
    if models and not models_err then
        print(string.format("Available models: %d", #models))
        for i, m in ipairs(models) do
            if i <= 5 then
                print(string.format("  - %s (%s)", m.name or m.id, m.id))
            end
        end
        if #models > 5 then
            print(string.format("  ... and %d more", #models - 5))
        end
    end

    -- Create a session with our custom tool
    local session, sess_err = client:create_session({
        tools               = { lookup_fact_tool },
        onPermissionRequest = permission_handler,
    })
    if not session then
        print("ERROR: Failed to create session: " .. tostring(sess_err))
        client:stop()
        os.exit(1)
    end
    print(string.format("Session created: %s", session.session_id))
    print()

    -- Subscribe to all events
    local unsubscribe = session:on(function(event)
        local event_type = event.type or "unknown"
        print(string.format("[Event] %s", event_type))
        if event.data and event.data.content then
            print(string.format("  Content: %s",
                tostring(event.data.content):sub(1, 200)))
        end
    end)

    -- Send a simple message
    print("Sending message: 'What is 2+2?'")
    local response1, err1 = session:send_and_wait({ prompt = "What is 2+2?" })
    if err1 then
        print("ERROR: " .. tostring(err1))
    elseif response1 then
        print(string.format("Response: %s",
            tostring(response1.data and response1.data.content or "(no content)")))
    end
    print()

    -- Send a message that exercises the custom tool
    print("Sending message: 'Use lookup_fact to tell me about lua'")
    local response2, err2 = session:send_and_wait({
        prompt = "Use the lookup_fact tool to tell me about 'lua'",
    })
    if err2 then
        print("ERROR: " .. tostring(err2))
    elseif response2 then
        print(string.format("Response: %s",
            tostring(response2.data and response2.data.content or "(no content)")))
    end
    print()

    -- Unsubscribe from events
    unsubscribe()

    -- Get conversation history
    local messages, msg_err = session:get_messages()
    if messages and not msg_err then
        print(string.format("Conversation history: %d events", #messages))
    end

    -- --- v2.0 Features ---

    -- Session Metadata
    local meta = client:get_session_metadata(session.session_id)
    if meta then
        print("Session ID: " .. meta.session_id)
    end

    -- Skills (uncomment to use)
    -- local skill_session = client:create_session({
    --     skill_directories = { "./skills" },
    --     include_sub_agent_streaming_events = true,
    -- })

    -- Clean up
    print("Destroying session...")
    session:destroy()

    print("Stopping client...")
    client:stop()

    print(string.rep("-", 50))
    print("Done!")
end

-- Run
main()
